pkg/apis/v1beta1/nodeclass.go

```diff
@@ -0,0 +1,278 @@
+/*
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1beta1
+
+import (
+	"fmt"
+
+	"github.com/mitchellh/hashstructure/v2"
+	"k8s.io/apimachinery/pkg/api/resource"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// NodeClassSpec is the top level specification for the AWS Karpenter Provider.
+// This will contain configuration necessary to launch instances in AWS.
+type NodeClassSpec struct {
+	// AMIFamily is the AMI family that instances use.
+	// +optional
+	AMIFamily *string `json:"amiFamily,omitempty"`
+	// UserData to be applied to the provisioned nodes.
+	// It must be in the appropriate format based on the AMIFamily in use. Karpenter will merge certain fields into
+	// this UserData to ensure nodes are being provisioned with the correct configuration.
+	// +optional
+	UserData *string `json:"userData,omitempty"`
+	// Role is the AWS identity that nodes use.
+	// +optional
+	Role *string `json:"role,omitempty"`
+	// SubnetSelectorTerms is a list of or subnet selector terms. The terms are ORed.
+	// +optional
+	SubnetSelectorTerms []SubnetSelectorTerm `json:"subnetSelectorTerms" hash:"ignore"`
```

Comment:
I kind of see this as akin to naming []foo as `fooSlice`. There are many types of selectors in k8s. 
```suggestion
	SubnetSelector []SubnetSelectorTerm `json:"subnetSelector" hash:"ignore"`
```

Reply:
Discussed offline. The consensus among us was that upstream K8s conveys ORed selectors through the word "terms"; therefore, we are going to propagate the word terms into our selector logic to convey the same thing.